File: example_classification/example_classification.cc
#include <opencv2/videoio.hpp>
#include <opencv2/imgproc.hpp>
#include <opencv2/highgui.hpp>
#include "tensorflow/core/framework/tensor.h"
#include "tensorflow/lite/interpreter.h"
#include "tensorflow/lite/kernels/register.h"
#include "tensorflow/lite/model.h"
#include "tensorflow/lite/tools/gen_op_registration.h"
#ifdef USE_GPU
#include "tensorflow/lite/delegates/gpu/delegate.h"
#endif // USE_GPU
#include <iostream>
#include <algorithm>
#include <string>
#include <fstream>
#include "absl/flags/flag.h"
#include "absl/flags/parse.h"

const std::string DEFAULT_MODEL = "./example_classification/mobilenet.tflite";
const std::string DEFAULT_IMAGE = "./images/goldfish.png";
const std::string DEFAULT_LABELS = "./example_classification/labels.txt";

ABSL_FLAG(std::string, model_file, DEFAULT_MODEL, "TFLite model to use for the inference; Note : input must have RGB channels");
ABSL_FLAG(std::string, image_file, DEFAULT_IMAGE, "Image file to run the inference on");
ABSL_FLAG(std::string, label_file, DEFAULT_LABELS, "File containing the labels for the classification");
ABSL_FLAG(int, number_of_inferences, 1, "Number of inference to run");

void ReadAndBindImageToTfInput(const char *image_file, float *input_tensor, TfLiteIntArray *dims)
{
    cv::Mat image = cv::imread(image_file);
    int image_width = image.size().width;
    int image_height = image.size().height;
    int square_dim = std::min(image_height, image_height);
    int delta_height = (image_height - square_dim) / 2;
    int delta_width = (image_width - square_dim) / 2;

    cv::Mat out_frame;
    // Crop the biggest centered square inside the image
    cv::resize(image(cv::Rect(delta_width, delta_height, square_dim, square_dim)), out_frame, cv::Size(dims->data[1], dims->data[2]));
    cv::Mat tensor_image(dims->data[1], dims->data[2], CV_32FC3, input_tensor);
    out_frame.convertTo(tensor_image, CV_32FC3);
    cv::cvtColor(tensor_image, tensor_image, cv::COLOR_BGR2RGB);
}

void ReadOutput(const char *label_file, float *output_tensor)
{
    std::ifstream in(label_file);
    std::string str;
    std::vector<std::string> labels;
    while (std::getline(in, str))
    {
        if (str.size() > 0)
            labels.push_back(str);
    }
    std::vector<std::pair<float, std::string>> ordered_scores;
    for (int i = 0; i < labels.size(); ++i)
    {
        float score = output_tensor[i];
        bool inserted = false;
        std::pair<float, std::string> current_pair{score, labels[i]};
        std::vector<std::pair<float, std::string>>::iterator it;
        for (it = ordered_scores.begin(); it != ordered_scores.end(); it++)
        {
            float score_it = (*it).first;
            if (score < score_it)
            {
                ordered_scores.insert(it, current_pair);
                inserted = true;
                break;
            }
        }
        if (!inserted)
            ordered_scores.emplace_back(current_pair);
    }

    std::reverse(ordered_scores.begin(), ordered_scores.end());

    std::cout << std::endl;
    std::cout << "Classification results : " << std::endl;
    for (int i = 0; i < 5; ++i)
    {

        std::cout << "Label : " << ordered_scores[i].second << " - Score : " << ordered_scores[i].first << std::endl;
    }
}

void run_inference(const char *image_file, const char *model_path, const char *label_file, int number_of_inferences)
{
    auto model = tflite::FlatBufferModel::BuildFromFile(model_path);

    if (!model)
    {
        std::cout << "Could not build model from : " << model_path << std::endl;
        return;
    }
    tflite::ops::builtin::BuiltinOpResolver op_resolver;
    std::unique_ptr<tflite::Interpreter> interpreter;
    tflite::InterpreterBuilder(*model, op_resolver)(&interpreter);

#ifdef USE_GPU
    auto *delegate = TfLiteGpuDelegateV2Create(nullptr);
    if (interpreter->ModifyGraphWithDelegate(delegate) != kTfLiteOk)
    {
        std::cout << "Could not setup GPU delegate" << std::endl;
        return;
    }
#endif // USE_GPU

    interpreter->AllocateTensors();
    auto input = interpreter->inputs()[0];

    TfLiteIntArray *dims = interpreter->tensor(input)->dims;

    // Read the image and emplace it in the model input
    ReadAndBindImageToTfInput(image_file, interpreter->typed_input_tensor<float>(0), dims);

    auto start = std::chrono::system_clock::now();
    // Run the inferences on the image
    for (int i = 0; i < number_of_inferences; ++i)
    {
        if (interpreter->Invoke() != kTfLiteOk)
        {
            std::cout << "Invoke failed" << std::endl;
            return;
        }
    }
    auto end = std::chrono::system_clock::now();
    std::chrono::duration<double> elapsed_seconds = end - start;
    std::cout << std::endl;
    std::cout << "Ran " << number_of_inferences << " inferences in " << elapsed_seconds.count() << std::endl
              << "Average time per inference : " << 1000 * elapsed_seconds.count() / number_of_inferences << " ms" << std::endl
              << "Average inference per seconds : " << number_of_inferences / elapsed_seconds.count() << std::endl;

    // Process the ouptput
    ReadOutput(label_file, interpreter->typed_output_tensor<float>(0));
}

int main(int argc, char **argv)
{
    /**
     * MUST HAVE THE 4 FOLLOWING ARGUMENTS : 
     * model_file : path to the tflite_model
     * label_file : path to the output labels
     * image_file : path to the image to run the inference on
    */
    absl::ParseCommandLine(argc, argv);
    std::string model_file = absl::GetFlag(FLAGS_model_file);
    std::string image_file = absl::GetFlag(FLAGS_image_file);
    std::string label_file = absl::GetFlag(FLAGS_label_file);
    int number_of_inferences = absl::GetFlag(FLAGS_number_of_inferences);

    std::cout
        << "Inference details : " << std::endl
        << "model_file : " << model_file << std::endl
        << "label_file : " << label_file << std::endl
        << "image_file : " << image_file << std::endl
        << "number_of_inference : " << number_of_inferences << std::endl;

    run_inference(image_file.c_str(), model_file.c_str(), label_file.c_str(), number_of_inferences);
}

File: example_classification/example_classification.py
import tensorflow as tf
import numpy as np
import time
import sys
import argparse
import cv2

def run_inference(model_file, label_file, image_file, number_of_inferences):

    interpreter = tf.lite.Interpreter(model_path=model_file)

    interpreter.allocate_tensors()

    input_details = interpreter.get_input_details()
    output_details = interpreter.get_output_details()

    input_img = cv2.cvtColor(cv2.imread(image_file), cv2.COLOR_BGR2RGB).astype(np.float32)
    image_height, image_width, _ = input_img.shape
    square_dim = min(image_height, image_width)
    delta_height = (image_height - square_dim) // 2
    delta_width = (image_width - square_dim) // 2
    cropped_img = input_img[delta_height:square_dim+delta_height, delta_width:square_dim+delta_width]

    input_shape = (input_details[0]['shape'][1], input_details[0]['shape'][2])
    resized_img = cv2.resize(cropped_img, input_shape)

    interpreter.set_tensor(input_details[0]['index'], [resized_img])

    start = time.time()
    for i in range(number_of_inferences):
        interpreter.invoke()
    end = time.time()

    elapsed_seconds = end - start
    print()
    print(f"Ran {number_of_inferences} inferences in {elapsed_seconds}")
    print(f"Average time per inference : {1000 * elapsed_seconds / number_of_inferences} ms")
    print(f"Average number of inference per seconds : {number_of_inferences / elapsed_seconds}")


    scores = interpreter.get_tensor(output_details[0]['index'])[0]

    with open(label_file, "r") as f:
        content = f.readlines()

    labels = [x.strip() for x in content] 



    score_labels = [(labels[i], scores[i]) for i in range(len(labels))]

    comp = lambda x : x[1]

    score_labels.sort(key = comp, reverse=True)
    print()
    print("Classification results : ")
    for i in range(5):
        print(f"Label : {score_labels[i][0]} - Score : {score_labels[i][1]}")

def main():
    parser = argparse.ArgumentParser()
    parser.add_argument("--model_file", default="./example_classification/mobilenet.tflite", help="TFLite model to use for the inference; Note : input must have RGB channels")
    parser.add_argument("--image_file", default="./images/goldfish.png", help="Image file to run the inference on")
    parser.add_argument("--label_file", default="./example_classification/labels.txt", help="File containing the labels for the classification")
    parser.add_argument("--number_of_inferences", default=1, help="number of inferences to run", type=int)
    args = parser.parse_args()
    print("Inference details : ")
    print(f"model_file : {args.model_file}")
    print(f"label_file : {args.label_file}")
    print(f"image_file : {args.image_file}")
    print(f"number_of_inference : {args.number_of_inferences}")
    run_inference(args.model_file, args.label_file, args.image_file, args.number_of_inferences)

if __name__ == "__main__":
    main()